Improve node draining

calvix opened the pull request:
towards https://github.com/giantswarm/giantswarm/issues/4942

* use EvictApi for draining fo pods
* execute eviction of each pod in a separate goroutine
* wait until the pod is terminated
* skip deleting of static critical pods (api, controller-manager and scheduler)

xh3b4sd commented on service/controller/v2/resource/node/create.go:
```diff
@@ -114,30 +123,58 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
-			}
+			// delete each pod in separate goroutine to avoid blocking
+			wg.Add(1)
+			go func(k8sClient kubernetes.Interface, pod v1.Pod) {
```
I would like to use https://godoc.org/golang.org/x/sync/errgroup for parallel execution. It is our standard way of doing it and has better error handling capabilities. Also note the variable scope binding in loops. This is very important. The code is broken right now already and only the last pod will be managed.

xh3b4sd commented on service/controller/v2/resource/node/create.go:
```diff
@@ -114,30 +123,58 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+				}
+				b := backoff.NewExponential(backoff.ShortMaxWait, backoff.ShortMaxInterval)
+				n := backoff.NewNotifier(r.logger, ctx)
+				err = backoff.RetryNotify(o, b, n)
```
I am not sure we should retry here, because we have the retry resource. @kopiczko what is your idea here? How should that be at best?

kopiczko replied:
I think short ResyncPeriod and aggressive cancelling on error. I like your suggestion about `errgroup`. I think nodeconfigs are short living objects so we can afford listing them often.

Blocking for 2 min (ShortMaxWait) doesn't make any sense to me. We can just set ResyncPeriod to 2min. Only backoff that would make sense here IMO is just `NewMaxRetries`. Something like `backoff.MaxRetries(3, 1*time.Second)` especially when this is done in concurrent fashion. We can spend like couple of seconds in the loop in the worst case scenario I think.

xh3b4sd commented on service/controller/v2/resource/node/create.go:
```diff
@@ -114,30 +123,58 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+			go func(k8sClient kubernetes.Interface, pod v1.Pod) {
+				defer wg.Done()
+				o := func() error {
+					return EvictPod(k8sClient, pod)
```
Please always do explicit error handling. This code hides the trace of the error now, which makes debugging harder.

xh3b4sd commented on service/controller/v2/resource/node/evict.go:
```diff
@@ -0,0 +1,61 @@
+			// pod is no longer in api, we can exit
+			return nil
+		} else if err != nil {
+			return err
```
Error masking is missing here.

xh3b4sd commented on service/controller/v2/resource/node/evict.go:
```diff
@@ -0,0 +1,61 @@
+	getOpts := apismetav1.GetOptions{}
+	// wait for successful termination
+	b := backoff.NewConstant(waitTimeout, waitInterval)
+	o := func() error {
```
With this we now do a retry with the retry resource, a retry with the drainer resource, and a retry with this code here. That does not make sense. We should only have one retry at one level.

xh3b4sd commented on service/controller/v2/resource/node/evict.go:
```diff
@@ -0,0 +1,61 @@
+		p, err := k8sClient.CoreV1().Pods(pod.Namespace).Get(pod.Name, getOpts)
+		if apierrors.IsNotFound(err) || (p != nil && p.ObjectMeta.UID != pod.ObjectMeta.UID) {
+			// pod is no longer in api, we can exit
+			return nil
```
I think the approach of blocking the resource here is wrong. We should aim for short running resources. I would suggest to initiate the eviction and cancel the reconciliation without setting the CR status to drained. In the next loop we should check if all pods are evicted and then set the CR status accordingly.

calvix replied:
WOuld it than make sense to introduce new state into CR? Or where would it be a good place to store information that we created eviction and now we are waiting for pod termination so we can act accordingly in the next loop.

xh3b4sd replied:
I do not know how the eviction API works yet. What happens when we evict multiple times? Does k8s provide information that we already evicted? Then we would not need to track such state by ourselves. We could maybe just evict again and act based on the response.

kopiczko replied:
@calvix can't you just check if the pod is evicted in the k8s api?

calvix replied:
Well, each time this loop run there will be a call for listing all pods on the node, but some pods like daemonSets can be scheduled on nodes that have `unschedulable` status.

If we ignore daemonSets than we don't have to save any status but just repeat the loop until no more pods are on node anymore. That could work.

In fact, `kubectl drain` also ignores daemonSet pods (by default it will exit with an error, but with arg `--ignore-daemonsets` it just ignore them)

kopiczko replied:
Are you afraid about daemonsets created by customers?

calvix replied:
Well, daemonSets are not usually  used for running workloads, but more for the operational thing ( ie calico, monitoring  and other.) So if some customer does some stupid thing and uses daemonSet for their web app then there is nothing we can do.

We can always say that we drain nodes same as `kubectl` drain does so we are synced with the community and it's their fault they run bad podController.

calvix commented:
@xh3b4sd if node resource is deprecated can I remove it from the v2 code? Or we need to somewhere write that its deprecated, as this is impossible to know from someone who does not work with it often.

xh3b4sd commented:
I did not yet manage to check that back but when you could quickly scan the provider operators, then there should not be any use of the node resources anymore so we then could drop the resource' functionality completely. But let's do this then in a separate PR and let's keep the resources, but drop the code for the logic.

calvix commented on service/controller/v2/resource/node/evict.go:
```diff
@@ -0,0 +1,61 @@
+		DeleteOptions: deleteOptions,
+	}
+	// evict pod
+	err := k8sClient.PolicyV1beta1().Evictions(eviction.Namespace).Evict(eviction)
```
There is not a big difference but  Eviction is respecting PodDistruptionBudget https://kubernetes.io/docs/tasks/run-application/configure-pdb/

That means that if you have deployments which specify that `maxUnavaiable is 1 then eviction will respect that and make sure that this is satisfied(IE:  only delete one pod and wait until the new one started on the new node and than delete another one). Delete would just kill all pods right away.

ALso in logs audit logos its more visible that eviction is some sort of process and delete is mostly used by the user.

calvix commented:
I rewrote the code and dropped the retry and wait logic.  Now it should not have a long-running loop and still be able to wait until all pods are evicted properly. It still needs some testing.

xh3b4sd commented on service/controller/v2/resource/drainer/create.go:
```diff
@@ -136,6 +136,22 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 		}
 
 		for _, p := range podList.Items {
+			if key.IsCriticalPod(p.Name) {
```
Really nice approach to only collect the relevant pods. Keeps the code below so much cleaner.

xh3b4sd commented on service/controller/v2/resource/drainer/create.go:
```diff
@@ -148,36 +164,37 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 
 	if len(customPods) > 0 {
-		r.logger.LogCtx(ctx, "level", "debug", "message", "deleting all pods running custom workloads")
+		r.logger.LogCtx(ctx, "level", "debug", "message", "evicting all pods running custom workloads")
```
```suggestion
		r.logger.LogCtx(ctx, "level", "debug", "message", "sending eviction to all pods running custom workloads")
```

xh3b4sd commented on service/controller/v2/resource/drainer/create.go:
```diff
@@ -148,36 +164,37 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "sent eviction to all pods running custom workloads")
 	} else {
-		r.logger.LogCtx(ctx, "level", "debug", "message", "no pods to be deleted running custom workloads")
+		r.logger.LogCtx(ctx, "level", "debug", "message", "no pods to be evicted running custom workloads")
```
```suggestion
		r.logger.LogCtx(ctx, "level", "debug", "message", "no pods running custom workloads to send evictions to")
```

xh3b4sd replied:
@rossf7 does that make sense?

xh3b4sd commented on service/controller/v2/resource/drainer/create.go:
```diff
@@ -148,36 +164,37 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 
 	if len(systemPods) > 0 {
-		r.logger.LogCtx(ctx, "level", "debug", "message", "deleting all pods running system workloads")
+		r.logger.LogCtx(ctx, "level", "debug", "message", "evicting all pods running system workloads")
```
```suggestion
		r.logger.LogCtx(ctx, "level", "debug", "message", "sending eviction to all pods running system workloads")
```

xh3b4sd commented on service/controller/v2/resource/drainer/create.go:
```diff
@@ -148,36 +164,37 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "sent eviction to all pods running system workloads")
 	} else {
-		r.logger.LogCtx(ctx, "level", "debug", "message", "no pods to be deleted running system workloads")
+		r.logger.LogCtx(ctx, "level", "debug", "message", "no pods to be evicted running system workloads")
```
```suggestion
		r.logger.LogCtx(ctx, "level", "debug", "message", "no pods running system workloads to send evictions to")
```

xh3b4sd commented on service/controller/v2/resource/drainer/create.go:
```diff
@@ -148,36 +164,37 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 	}
 
-	{
+	// when all pods are evicted from node, set status to drained
```
```suggestion
	// When all pods are evicted from the tenant node, set the CR status to drained.
```

xh3b4sd commented on service/controller/v2/resource/drainer/evict.go:
```diff
@@ -0,0 +1,34 @@
+)
+
+// evict pod from node
+func EvictPod(k8sClient kubernetes.Interface, pod v1.Pod) error {
```
```suggestion
func evictPod(k8sClient kubernetes.Interface, pod v1.Pod) error {
```

xh3b4sd commented on service/controller/v2/resource/drainer/evict.go:
```diff
@@ -0,0 +1,34 @@
+	"k8s.io/client-go/kubernetes"
+)
+
+// evict pod from node
```
```suggestion
```

xh3b4sd replied:
Let's drop this or make it compliant with godoc standards.

xh3b4sd commented on service/controller/v2/resource/drainer/evict.go:
```diff
@@ -0,0 +1,34 @@
+		},
+		DeleteOptions: deleteOptions,
+	}
+	// evict pod
```
```suggestion
```

xh3b4sd commented on service/controller/v2/resource/drainer/create.go:
```diff
@@ -148,36 +164,37 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 
 		for _, p := range systemPods {
-			err := k8sClient.CoreV1().Pods(p.GetNamespace()).Delete(p.GetName(), &metav1.DeleteOptions{})
+			err := EvictPod(k8sClient, p)
```
```suggestion
			err := evictPod(k8sClient, p)
```

xh3b4sd commented on service/controller/v2/resource/drainer/create.go:
```diff
@@ -148,36 +164,37 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 
 		for _, p := range customPods {
-			err := k8sClient.CoreV1().Pods(p.GetNamespace()).Delete(p.GetName(), &metav1.DeleteOptions{})
+			err := EvictPod(k8sClient, p)
```
```suggestion
			err := evictPod(k8sClient, p)
```

xh3b4sd commented on service/controller/v2/resource/drainer/evict.go:
```diff
@@ -0,0 +1,34 @@
+
+// evict pod from node
+func EvictPod(k8sClient kubernetes.Interface, pod v1.Pod) error {
+	var deleteGracePeriod int64 = 60
```
I need a refresher here. How does the grace period behave with the client? Will the eviction be scheduled and the k8s client returns immediately or will the client when calling `Evict` block for a minute now? When it is the first case, this is fine.

calvix replied:
it will returns immediately, it does not block

calvix commented:
It looks like it works as expected on first try, did not expect that :p

```
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/tenantcluster/tenantcluster.go:114","cluster":"wdj8s","level":"debug","message":"creating K8s client for the tenant cluster","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:00.866788+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/tenantcluster/tenantcluster.go:136","cluster":"wdj8s","level":"debug","message":"looking for certificates for the tenant cluster","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:00.866884+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/tenantcluster/tenantcluster.go:148","cluster":"wdj8s","level":"debug","message":"found certificates for the tenant cluster","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:00.869971+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/operatorkit/client/k8srestconfig/k8s_rest_config.go:134","level":"debug","message":"creating out-cluster config","time":"2019-01-21T16:02:00.870015+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/tenantcluster/tenantcluster.go:126","cluster":"wdj8s","level":"debug","message":"created K8s client for the tenant cluster","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:00.87081+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:83","cluster":"wdj8s","level":"debug","message":"cordoning guest cluster node","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:00.87089+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:119","cluster":"wdj8s","level":"debug","message":"cordoned guest cluster node","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:04.508073+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:125","cluster":"wdj8s","level":"debug","message":"looking for all pods running on the guest cluster node","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:04.508148+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:162","cluster":"wdj8s","level":"debug","message":"found 5 pods running custom workloads","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:04.623899+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:163","cluster":"wdj8s","level":"debug","message":"found 3 pods running system workloads","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:04.624+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:167","cluster":"wdj8s","level":"debug","message":"sending eviction to all pods running custom workloads","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:04.624028+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:176","cluster":"wdj8s","level":"debug","message":"sent eviction to all pods running custom workloads","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:05.016112+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:194","cluster":"wdj8s","level":"debug","message":"no pods running system workloads to send evictions to","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:05.01618+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/tenantcluster/tenantcluster.go:114","cluster":"wdj8s","level":"debug","message":"creating K8s client for the tenant cluster","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:05.016248+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/tenantcluster/tenantcluster.go:136","cluster":"wdj8s","level":"debug","message":"looking for certificates for the tenant cluster","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:05.016273+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/tenantcluster/tenantcluster.go:148","cluster":"wdj8s","level":"debug","message":"found certificates for the tenant cluster","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:05.022639+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/operatorkit/client/k8srestconfig/k8s_rest_config.go:134","level":"debug","message":"creating out-cluster config","time":"2019-01-21T16:02:05.022678+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/tenantcluster/tenantcluster.go:126","cluster":"wdj8s","level":"debug","message":"created K8s client for the tenant cluster","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:05.023284+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:83","cluster":"wdj8s","level":"debug","message":"cordoning guest cluster node","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:05.02335+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:119","cluster":"wdj8s","level":"debug","message":"cordoned guest cluster node","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:05.030732+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:125","cluster":"wdj8s","level":"debug","message":"looking for all pods running on the guest cluster node","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:05.030772+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:162","cluster":"wdj8s","level":"debug","message":"found 5 pods running custom workloads","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:05.059522+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:163","cluster":"wdj8s","level":"debug","message":"found 3 pods running system workloads","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:05.059567+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:167","cluster":"wdj8s","level":"debug","message":"sending eviction to all pods running custom workloads","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:05.0596+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:176","cluster":"wdj8s","level":"debug","message":"sent eviction to all pods running custom workloads","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:05.296658+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:194","cluster":"wdj8s","level":"debug","message":"no pods running system workloads to send evictions to","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:02:05.296745+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/exporterkit/collector/set.go:73","level":"debug","message":"collecting metrics","time":"2019-01-21T16:02:05.984343+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/exporterkit/collector/set.go:95","level":"debug","message":"collected metrics","time":"2019-01-21T16:02:06.9865+00:00"}
---
^[[A{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/exporterkit/collector/set.go:73","level":"debug","message":"collecting metrics","time":"2019-01-21T16:03:05.980602+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/exporterkit/collector/set.go:95","level":"debug","message":"collected metrics","time":"2019-01-21T16:03:06.987791+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/tenantcluster/tenantcluster.go:114","cluster":"wdj8s","level":"debug","message":"creating K8s client for the tenant cluster","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:03:57.046651+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/tenantcluster/tenantcluster.go:136","cluster":"wdj8s","level":"debug","message":"looking for certificates for the tenant cluster","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:03:57.046841+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/tenantcluster/tenantcluster.go:148","cluster":"wdj8s","level":"debug","message":"found certificates for the tenant cluster","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:03:57.0503+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/operatorkit/client/k8srestconfig/k8s_rest_config.go:134","level":"debug","message":"creating out-cluster config","time":"2019-01-21T16:03:57.050368+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/tenantcluster/tenantcluster.go:126","cluster":"wdj8s","level":"debug","message":"created K8s client for the tenant cluster","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:03:57.050905+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:83","cluster":"wdj8s","level":"debug","message":"cordoning guest cluster node","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:03:57.050956+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:119","cluster":"wdj8s","level":"debug","message":"cordoned guest cluster node","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:03:57.057853+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:125","cluster":"wdj8s","level":"debug","message":"looking for all pods running on the guest cluster node","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:03:57.057884+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:162","cluster":"wdj8s","level":"debug","message":"found 0 pods running custom workloads","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:03:57.067634+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:163","cluster":"wdj8s","level":"debug","message":"found 3 pods running system workloads","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:03:57.067668+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:178","cluster":"wdj8s","level":"debug","message":"no pods running custom workloads to send evictions to","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:03:57.067684+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:183","cluster":"wdj8s","level":"debug","message":"sending eviction to all pods running system workloads","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:03:57.067697+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:192","cluster":"wdj8s","level":"debug","message":"sent eviction to all pods running system workloads","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:03:57.137801+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/exporterkit/collector/set.go:73","level":"debug","message":"collecting metrics","time":"2019-01-21T16:04:05.980508+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/exporterkit/collector/set.go:95","level":"debug","message":"collected metrics","time":"2019-01-21T16:04:06.991204+00:00"}
---
---
---
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/exporterkit/collector/set.go:73","level":"debug","message":"collecting metrics","time":"2019-01-21T16:05:05.981359+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/exporterkit/collector/set.go:95","level":"debug","message":"collected metrics","time":"2019-01-21T16:05:06.983721+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/tenantcluster/tenantcluster.go:114","cluster":"wdj8s","level":"debug","message":"creating K8s client for the tenant cluster","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:05:58.053339+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/tenantcluster/tenantcluster.go:136","cluster":"wdj8s","level":"debug","message":"looking for certificates for the tenant cluster","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:05:58.053452+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/tenantcluster/tenantcluster.go:148","cluster":"wdj8s","level":"debug","message":"found certificates for the tenant cluster","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:05:58.056019+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/operatorkit/client/k8srestconfig/k8s_rest_config.go:134","level":"debug","message":"creating out-cluster config","time":"2019-01-21T16:05:58.056157+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/tenantcluster/tenantcluster.go:126","cluster":"wdj8s","level":"debug","message":"created K8s client for the tenant cluster","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:05:58.056939+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:83","cluster":"wdj8s","level":"debug","message":"cordoning guest cluster node","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:05:58.05709+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:119","cluster":"wdj8s","level":"debug","message":"cordoned guest cluster node","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:05:58.062884+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:125","cluster":"wdj8s","level":"debug","message":"looking for all pods running on the guest cluster node","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:05:58.062931+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:162","cluster":"wdj8s","level":"debug","message":"found 0 pods running custom workloads","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:05:58.071545+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:163","cluster":"wdj8s","level":"debug","message":"found 0 pods running system workloads","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:05:58.071647+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:178","cluster":"wdj8s","level":"debug","message":"no pods running custom workloads to send evictions to","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:05:58.071684+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:194","cluster":"wdj8s","level":"debug","message":"no pods running system workloads to send evictions to","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:05:58.071753+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:199","cluster":"wdj8s","level":"debug","message":"setting drainer config status of node in guest cluster 'wdj8s' to drained condition","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:05:58.071797+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:208","cluster":"wdj8s","level":"debug","message":"set drainer config status of node in guest cluster 'wdj8s' to drained condition","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:05:58.076508+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:35","cluster":"wdj8s","level":"debug","message":"drainer config status has drained condition","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:05:58.076581+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:36","cluster":"wdj8s","level":"debug","message":"canceling resource","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:05:58.076636+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/exporterkit/collector/set.go:73","level":"debug","message":"collecting metrics","time":"2019-01-21T16:06:05.980396+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/exporterkit/collector/set.go:95","level":"debug","message":"collected metrics","time":"2019-01-21T16:06:06.983659+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/exporterkit/collector/set.go:73","level":"debug","message":"collecting metrics","time":"2019-01-21T16:07:05.980919+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/exporterkit/collector/set.go:95","level":"debug","message":"collected metrics","time":"2019-01-21T16:07:06.987593+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:35","cluster":"wdj8s","level":"debug","message":"drainer config status has drained condition","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:07:59.05555+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/create.go:36","cluster":"wdj8s","level":"debug","message":"canceling resource","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:07:59.055648+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/tenantcluster/tenantcluster.go:114","cluster":"wdj8s","level":"debug","message":"creating K8s client for the tenant cluster","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:08:00.140376+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/tenantcluster/tenantcluster.go:136","cluster":"wdj8s","level":"debug","message":"looking for certificates for the tenant cluster","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:08:00.140503+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/tenantcluster/tenantcluster.go:148","cluster":"wdj8s","level":"debug","message":"found certificates for the tenant cluster","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:08:00.143593+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/operatorkit/client/k8srestconfig/k8s_rest_config.go:134","level":"debug","message":"creating out-cluster config","time":"2019-01-21T16:08:00.143644+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/tenantcluster/tenantcluster.go:126","cluster":"wdj8s","level":"debug","message":"created K8s client for the tenant cluster","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:08:00.144311+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/delete.go:38","cluster":"wdj8s","level":"debug","message":"deleting guest cluster node from Kubernetes API","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:08:00.144361+00:00"}
{"caller":"github.com/giantswarm/node-operator/service/controller/v2/resource/drainer/delete.go:59","cluster":"wdj8s","level":"debug","message":"deleted guest cluster node from Kubernetes API","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:08:00.155296+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/operatorkit/controller/finalizer.go:146","cluster":"wdj8s","level":"debug","message":"removing finalizer 'operatorkit.giantswarm.io/node-operator'","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:08:00.155347+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/operatorkit/controller/finalizer.go:187","cluster":"wdj8s","level":"debug","message":"removed finalizer 'operatorkit.giantswarm.io/node-operator'","node":"wdj8s-worker-000001","resource":"drainerv2","time":"2019-01-21T16:08:00.162535+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/operatorkit/controller/controller.go:220","cluster":"wdj8s","level":"debug","message":"did not find any finalizer","node":"wdj8s-worker-000001","time":"2019-01-21T16:08:00.162607+00:00"}
{"caller":"github.com/giantswarm/node-operator/vendor/github.com/giantswarm/operatorkit/controller/controller.go:221","cluster":"wdj8s","level":"debug","message":"canceling reconciliation","node":"wdj8s-worker-000001","time":"2019-01-21T16:08:00.162646+00:00"}

```

calvix commented:
So yeah RFR

kopiczko commented on service/controller/v2/key/key.go:
```diff
@@ -21,6 +25,30 @@ func ClusterIDFromDrainerConfig(drainerConfig v1alpha1.DrainerConfig) string {
+	return r
+}
+
+func IsEvicted(pod v1.Pod) bool {
```
```suggestion
func IsEvictedPod(pod v1.Pod) bool {
```

kopiczko commented on service/controller/v2/resource/drainer/create.go:
```diff
@@ -136,6 +136,22 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+				// we are aligning here with community as 'kubectl drain' also ignore them
+				continue
+			}
+			if key.IsEvicted(p) {
```
```suggestion
			if key.IsEvictedPod(p) {
```

calvix commented:
tested few times, merging

xh3b4sd commented:
🎉
